buffer/src/main/java/io/netty/buffer/api/Buffer.java

```diff
@@ -0,0 +1,685 @@
+     * {@linkplain #readerOffset() read offset} on this buffer, nor on the destination buffer.
+     * <p>
+     * The read and write offsets of the destination buffer are also ignored, and do not influence {@code destPos}
+     * or {@code length}.
+     *
+     * @param srcPos The byte offset into this buffer wherefrom the copying should start; the byte at this offset in
+     *              this buffer will be copied to the {@code destPos} index in the {@code dest} array.
+     * @param dest The destination buffer.
+     * @param destPos The index into the {@code dest} array wherefrom the copying should start.
+     * @param length The number of bytes to copy.
+     * @throws NullPointerException if the destination array is null.
+     * @throws IndexOutOfBoundsException if the source or destination positions, or the length, are negative,
+     * or if the resulting end positions reaches beyond the end of either this buffer, or the destination array.
+     */
+    void copyInto(int srcPos, Buffer dest, int destPos, int length);
+
+    /**
+     * Writes into this buffer, all the readable bytes from the given buffer.
+     * This updates the {@linkplain #writerOffset() write offset} of this buffer, and the
+     * {@linkplain #readerOffset() reader offset} of the given buffer.
+     *
+     * @param source The buffer to read from.
+     * @return This buffer.
+     */
+    default Buffer writeBytes(Buffer source) {
+        int size = source.readableBytes();
+        int woff = writerOffset();
+        writerOffset(woff + size);
+        source.copyInto(source.readerOffset(), this, woff, size);
+        source.readerOffset(source.readerOffset() + size);
+        return this;
+    }
+
+    /**
+     * Writes into this buffer, all the bytes from the given byte array.
+     * This updates the {@linkplain #writerOffset() write offset} of this buffer by the length of the array.
+     *
+     * @param source The byte array to read from.
+     * @return This buffer.
+     */
+    default Buffer writeBytes(byte[] source) {
+        int size = source.length;
+        int woff = writerOffset();
+        writerOffset(woff + size);
+        for (int i = 0; i < size; i++) {
+            setByte(woff + i, source[i]);
+        }
+        return this;
+    }
+
+    /**
+     * Resets the {@linkplain #readerOffset() read offset} and the {@linkplain #writerOffset() write offset} on this
+     * buffer to their initial values.
+     */
+    default Buffer reset() {
+        readerOffset(0);
+        writerOffset(0);
+        return this;
+    }
+
+    /**
+     * Opens a cursor to iterate the readable bytes of this buffer. The {@linkplain #readerOffset() reader offset} and
+     * {@linkplain #writerOffset() writer offset} are not modified by the cursor.
+     * <p>
+     * Care should be taken to ensure that the buffer's lifetime extends beyond the cursor and the iteration, and that
```

Comment:
Not without a synchronising action on every access. Same problem as closing a buffer that is being accessed concurrently.

Reply:
ok I should have explained more and perhaps `CME` isn't the best example. My understanding here is that there are two ways the reader/writer offset can be modified during iteration:

1. Concurrently modifying the buffer in a different thread.
2. Modifying the buffer by interacting with the buffer apart from the methods provided by the cursor, eg: writing to the buffer while a cursor is open. 

I was suggesting to avoid (2) since (1) anyways is not supported by the buffer

Reply:
I'm still not convinced. A buffer is not the same as an optimised `List<Byte>` in my mind. It would prevent you from using a cursor to make a read+write loop that operates on the same buffer, and I've seen use cases like that before.